### ISSUE TITLE
Git commit messages and pull requests

### ISSUE THREAD

**Issue by warpech** [Starcounter/Guidelines#13]
Hi apps team, I would like to propose some guidelines regarding commit messages. 

Maybe there is something that we can improve here.

Many times when I look at Git history I cannot tell why a change was made. This also applies to myself. Today I was not able to tell why did I have to add `<form>` element in this commit: https://github.com/StarcounterApps/SignIn/commit/89361a04400c0874c012a8cdeb62b5b4280f8299#diff-811c3b3ab68899762803273081dc3075R6

The below is just a proposal. Anyone cares to share their opinion?

## Commit messages

I think that a nice commit message is one that explains what have you changed and why.

[Here is a great short article](http://rickardlindberg.me/writing/problem-in-commit-message/) that proposes how to write commit messages according to a template:

```
Problem: {What is the problem?}

Solution: {How does the change in this commit fix the problem?}
```

Few longer write ups about great commit messages are here: 

- https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/process/submitting-patches.rst
- https://github.com/conventional-changelog/conventional-changelog/blob/a5505865ff3dd710cf757f50530e73ef0ca641da/conventions/angular.md
- https://robots.thoughtbot.com/5-useful-tips-for-a-better-commit-message
- https://github.com/erlang/otp/wiki/Writing-good-commit-messages

## Issue references

Another rule is that the Git commit should always include a reference to the relevant GitHub issue in form `#{number of the issue}`. GitHub recognizes this pattern and shows a cross-reference.

## Changing git message after committing

If you have not yet pushed your commits, you can still edit the commit messages.

To edit the commit message of the last commit, you can simply change it using `git commit --amend`. Sourcetree and other Git GUIs also give an option to amend a previous commit.

To edit commit messages of any commit you can use Git interactive rebase: https://git-scm.com/book/id/v2/Git-Tools-Rewriting-History

Remember though to never change commits that have been pushed to upstream!

## Pull requests

A pull request is just a reference to a branch that contains a bunch of commits. If your commits adhere to the above rules, then the description of the pull request can be even empty!

Chances are though, that after pushing your commits you want to add some details for the reviewer. Then the description of the pull request is a perfect place for it.

**Comment by Mackiovello** [comment]
> Another rule is that the Git commit should always include a reference to the relevant GitHub issue in form #{number of the issue}. GitHub recognizes this pattern and shows a cross-reference.

This one is brilliant! Tracking commits together with issues is super convenient.

How is it with naming of branches? I personally find that quite challenging. What are your thoughts on that?  

Do you have any thoughts on [closing issues via commits](https://help.github.com/articles/closing-issues-via-commit-messages/)? It could prevent solved issues from being left open.

I think all the things you bring up are highly reasonable. The tricky part is to make everyone use them.

**Comment by warpech** [comment]
> How is it with naming of branches? I personally find that quite challenging. What are your thoughts on that?

So far everyone seems to follow their own convention. Since feature branches are short-lived, I don't care much about the naming convention for the branches. 

Just one remark - putting the issue number in the branch name is usually helpful.

> Do you have any thoughts on closing issues via commits? It could prevent solved issues from being left open.

Actually usually it rarely works in our GitHub usage model. GitHub closes issues via commit when the commit is merged to the `master` branch. In our case we usually close the issues manually earlier, when they are merged to the `develop` branch.

Any suggestions here?

> I think all the things you bring up are highly reasonable. The tricky part is to make everyone use them.

I think that anyone who is convinced about the gains we get from this will use it.

Interestingly, Angular.js has some strict (I think too strict) commit message guidelines: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit

**Comment by warpech** [comment]
Implemented in #43

**Comment by warpech** [comment]
Closing as done.
